lib/src/test/java/io/ably/lib/http/HttpTest.java

```diff
@@ -0,0 +1,447 @@
+     * <p>
+     * Validates {@code Http} is using its host (non-fallback-host) first
+     * when a consecutive call happens
+     * </p>
+     * <p>
+     * Spec: -
+     * </p>
+     *
+     * @throws Exception
+     */
+    @Test
+    public void http_execute_consecutivecall() throws Exception {
+        Http http = Mockito.spy(new Http(new ClientOptions(), null));
+
+        String hostExpected = "non.fallback.host";
+        http.setHost(hostExpected);
+        ArgumentCaptor<URL> argument = ArgumentCaptor.forClass(URL.class);
+
+        /* Partially mock http */
+        doAnswer(new GrumpyAnswer(1, new Http.HostFailedException(new Exception()), null))
+                .when(http).httpExecute(argument.capture(), anyString(), aryEq(new Param[0]),
+                any(Http.RequestBody.class), anyBoolean(), any(Http.ResponseHandler.class));
+
+        http.ablyHttpExecute("", "", new Param[0], new Param[0],
+                Mockito.mock(Http.RequestBody.class), Mockito.mock(Http.ResponseHandler.class));
+
+        /* Verify there was a fallback with first call */
+        assertThat(argument.getValue().getHost().matches(PATTERN_HOST_FALLBACK), is(true));
+
+        /* Update behavior to perform a call without a fallback */
+        argument = ArgumentCaptor.forClass(URL.class);
+        doReturn(null)
+                .when(http).httpExecute(argument.capture(), anyString(), aryEq(new Param[0]),
+                any(Http.RequestBody.class), anyBoolean(), any(Http.ResponseHandler.class));
+
+        http.ablyHttpExecute("", "", new Param[0], new Param[0],
+                Mockito.mock(Http.RequestBody.class), Mockito.mock(Http.ResponseHandler.class));
+
+        /* Verify second call was called with http host */
+        assertThat(argument.getValue().getHost().equals(hostExpected), is(true));
+    }
+
+    /**
+     * <p>
+     * Validates {@code Http} is throwing an exception,
+     * when connection to host failed more than allowed count ({@code Defaults.HTTP_MAX_RETRY_COUNT})
+     * </p>
+     * <p>
+     * Spec: -
+     * </p>
+     *
+     * @throws Exception
+     */
+    @Test
+    public void http_execute_excessivefallback () throws AblyException {
+        ClientOptions options = new ClientOptions();
+        Http http = Mockito.spy(new Http(options, null));
+
+        ArgumentCaptor<URL> argument = ArgumentCaptor.forClass(URL.class);
+        int excessiveFallbackCount = options.httpMaxRetryCount + 1;
+
+        /* Partially mock http */
+        doAnswer(new GrumpyAnswer(excessiveFallbackCount, new Http.HostFailedException(new Exception()), ""))
+                .when(http).httpExecute(argument.capture(), anyString(), aryEq(new Param[0]),
+                any(Http.RequestBody.class), anyBoolean(), any(Http.ResponseHandler.class));
+
+
+        /* Verify
+         *   - ably exception with 404 status code is thrown
+         */
+        ErrorInfo expectedErrorInfo = new ErrorInfo("Connection failed; no host available", 404, 80000);
```

Comment:
This is testing behavior when client is failed more than (httpMaxRetryCount + 1). e.g., Assume httpMaxRetryCount is 1

  - Execute a request to host A. Let's assume it is failed
  - Fallback to host B. Let's assume it is also failed

Now can't do fallback and try again any more because of httpMaxRetry count constraint. Therefore, we are throwing an exception with 404